### ISSUE TITLE
Python module, Java package name collision

### ISSUE THREAD

**Issue by nullterminated** [jython/jython#64]
Trying to use a java package like io.vertx in a jython project results in errors when using the io python module like

jython-slim-2.7.2b2.jar/Lib/gzip.py", line 36, in <module>
AttributeError: 'javapackage' object has no attribute 'BufferedIOBase'

I've tracked the error down to the org.python.core.imp class method import_next which calls JavaImportHelper.tryAddPackage who calls JavaImportHelper.buildLoadedPackages. Since io is returned as a java package, the io module cannot be used.

I'm not sure how the team would like to fix the issue, but it presents a real challenge in using jython with a number of libraries.

**Comment by nullterminated** [comment]
I created a small [patch](https://github.com/nullterminated/jython/commit/1ddfa95f4cce975bed18409a60860fcc7d3797d2) that works for me. It reads a list of packages to omit from a system property. If you find this approach agreeable I will make a pull request.

**Comment by XenoAmess** [comment]
+1 for this...

**Comment by XenoAmess** [comment]
@Stewori
Hi. Any people interested in fixing this issue?
Many thanks as one of my repos need this.

**Comment by Stewori** [comment]
@XenoAmess Sorry, I can't (afford to) spend time on Jython development these days. I'm just helping to triage issues from time to time. If you have a fix, however, I would offer to review and possibly merge an appropriate PR.

**Comment by Stewori** [comment]
Just noticed that @nullterminated offered a patch. @XenoAmess does that patch work for you?

**Comment by Stewori** [comment]
I had a look at the patch and it introduces a new system property. I'm not sure whether this is the ideal way to fix it. I would say the right way to solve this is simply not to use `io` as a toplevel package name. This is a name conflict like any other. You would not put a new Python module under a toplevel package named `io` if this was all in pure Python, would you?
That said, maybe you can control the resolution order by modifying the order of `sys.path`.

**Comment by XenoAmess** [comment]
> You would not put a new Python module under a toplevel package named `io` if this was all in pure Python, would you?

correct, but Jython is a mixture of java and python, and in java, we usually use domain's reverse as ancesstor package name.

for example, I own domain xenoamess.com, so my packages in my repo always be under package *com.xenoamess*

And there be some kind of domains who ends with **.io** ...

(sigh)

**Comment by XenoAmess** [comment]
https://sg.godaddy.com/domainsearch/find?checkAvail=1&tmskey=1dom_03_consolidated&domainToCheck=xenoamess.io
for example...

**Comment by Stewori** [comment]
(was writing this in parallel...)

Okay, I  had a look at https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html. That states that

> Companies use their reversed Internet domain name to begin their package names—for example, com.example.mypackage for a package named mypackage created by a programmer at example.com.

So, is the actual issue that nowadays `.io` is a popular toplevel domain? Indeed, the webpage of `io.vertx` is `vertx.io`. Maybe the omit list is not such a bad idea to deal with this. @jeff5 what do you think?

Someone (not me. @XenoAmess ? @nullterminated ?) please do/check (if @jeff5 agrees):

- does the patch still allow to import (in Python code) the `io` module, and also `io.vertx`? If not, I'd suggest to make the latter available as `io_.vertx`, which would be in line with the advice from https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html.
- suggest a proper name for the new property. `"org.python.core.JavaImportHelper.omitPackages"` is too long and does not really cut it. It should be expressive and in line with existing Jython-specific property names (see [RegistryKey.java](https://github.com/jython/jython/blob/master/src/org/python/core/RegistryKey.java)).
- turn the patch into a PR, including the following steps:
- add an entry+doc for the new system property in [RegistryKey.java](https://github.com/jython/jython/blob/master/src/org/python/core/RegistryKey.java)
- adjust the patch to be robust against whitespaces (e.g. if someone writes `"io, video"` instead of `"io,video"`)
- make an entry in `NEWS`
- if not yet done, sign contributor agreement and add yourself to `ACKNOWLEDGMENTS`.
- check if there is some other place (including Jython website repo) where recognized system properties are listed or documented and possibly add the new property accordingly.

**Comment by XenoAmess** [comment]
@nullterminated please do it. many thanks!

**Comment by jeff5** [comment]
Yes, `io` is a popular TLD. It has fallen into commercial hands, its value having been underestimated by the government that rightfully owns it (mine, as it turns out).

I second @Stewori in suggesting we need a more succinct name. (I appreciate the design idea, but we could easily rename a class and still need the option.) And thanks for the pointer to `RegistryKey.java`, which reminds us there are `python.modules`, `python.packages` and `python.import` that look like relevant roots. How you use this ought to be documented there.

I don't like the default hard-coded in actual code at all. The default in code ought to be empty (current behaviour) and if there is a non-blank default it should be in the [registry](https://github.com/jython/jython/blob/master/registry). Although maybe it should only be a commented-out entry even there,

Values are read from that file, and other sources, into `PySystemState.registry` and then referenced from there across the codebase. `PySystemState.registry` is almost always the same as System.getProperties(), but the process for creating and filling it so convoluted its not a safe assumption.

`import` is a bit of a mess in Jython. It looks like it will do exactly what you assume it should, but I concluded that what you expect isn't strictly possible. A Java package and a Python module are similar, but not similar enough to neglect the difference.

I don't think I've really understood what effect this option has. I think the problem is that Java packages like io.vertx take the name, so it can't then be a Python module. Is that only when the path is pre-scanned into the package cache? And the solution stops the io TLD being cached visible to Python?

**Comment by XenoAmess** [comment]
Bad news: I tried this patch and build a jar and managed to replace it into my gradle build chain.

It doesn't work, at least in my situation...

**Comment by XenoAmess** [comment]
> I think the problem is that Java packages like io.vertx take the name, so it can't then be a Python module.

yep you are right, that is what cause the problem, but I have no idea how to fix it, and I see this issue and thought it could work.

...actually not.

**Comment by Stewori** [comment]
What if we generally made java packages that clash with a python module available with a trailing underscore? E.g. 'io_' No system property needed at all. Much like the java tutorial suggests if the package name clashes with a reserved name.

**Comment by XenoAmess** [comment]
> What if we generally made java packages that clash with a python module available with a trailing underscore? E.g. 'io_' No system property needed at all. Much like the java tutorial suggests if the package name clashes with a reserved name.

the only problem is, when we create a lib using jython, we should not know what packages will collide on the users' runtime...

so it be hard to determin where to add the score when writing the lib.

besides, score itself is legal in java package name, means there maybe some crazy people make package name "io_" (although mentally normal people would never do things like this)

**Comment by Stewori** [comment]
Can you describe in what way your setting differs from the original situation posted by the OP? IIUC there is some difference that causes the proposed patch to fail for you, right?

**Comment by jeff5** [comment]
I'm not sure that's how Stefan meant it ... but maybe. When wrestling with this before, I decided that what I wanted (or wanted to try, anyway) was *not* to distinguish Java and Python packages. So `io` would just be a package.

It's either that or fully distinguish them so they have separate namespaces, but then the Java packages have to be within a top level space `JARS.io.vertx` (or something).

There are other places where people (sensibly, I think) expect to put both Python and Java modules in the same namespace. And it almost works, but not quite, and you get to choose between two bugs, but you can't fix both. 

But I wasn't going to try this in Jython 2.

**Comment by Stewori** [comment]
> I'm not sure that's how Stefan meant it ... but maybe

I was just thinking loudly. What about the following resolution:

Whenever a Java package e.g. `io.vertx` is loaded while a Python module `io` already exists, the Java package is just added as `vertx` to the Python module's module dict. Then it would be available under its original fully qualified name.
Since the second level is a project name or company name, it should be rather unique and collisions on that level would be rare (in such rare cases, I suggest that an idiomatic alias name would be defined. Or it's simply an import error). If the Java package exists first, it would be replaced by the Python module and its second level would be added to the module dict. Same outcome.

That's just an idea. However, it still puzzles me why the patch works for @nullterminated  but not for @XenoAmess . No proper solution can be found before there is an explanation for that.

**Comment by jeff5** [comment]
@Stewori : That's exactly what I meant by "*not* to distinguish Java and Python packages", but explained better.

I assumed we would do that all down the hierarchy. t would deals the same way with potential collisions down the hierarchy. We would make `PyModule` and `PyJavaPackage` into one class.

There are subtleties to deal with about how lookup and directory works: what things should be in the dictionary of a module from the start, and which get added only by explicit import. The natural/feasible behaviour is different between Java and Python. (I know I haven't thought this through properly.) I'd probably remove the package cache as we know it.

**Comment by XenoAmess** [comment]
> > I'm not sure that's how Stefan meant it ... but maybe
> 
> I was just thinking loudly. What about the following resolution:
> 
> Whenever a Java package e.g. `io.vertx` is loaded while a Python module `io` already exists, the Java package is just added as `vertx` to the Python module's module dict. Then it would be available under its original fully qualified name. Since the second level is a project name or company name, it should be rather unique and collisions on that level would be rare (in such rare cases, I suggest that an idiomatic alias name would be defined. Or it's simply an import error). If the Java package exists first, it would be replaced by the Python module and its second level would be added to the module dict. Same outcome.
> 
> That's just an idea. However, it still puzzles me why the patch works for @nullterminated but not for @XenoAmess . No proper solution can be found before there is an explanation for that.

@Stewori

I make this demo for showing the problem I meet.

https://github.com/xenoamess-demo/demo_krakatau_jython_debug_1

the Krakatau_java_wrapper sources are at

https://github.com/XenoAmess/Krakatau_java_wrapper

the real world uscase I faced is complicated(as that is a jetbrains-idea plugin), so I tried to simplify it to demo_krakatau_jython_debug_1 for you.

If you find anything I did wrong please just tell me. I'm new to jython.

Hope this can help your investigation

**Comment by XenoAmess** [comment]
Hi.
Is there any news about this?
@Stewori @jeff5

**Comment by Stewori** [comment]
Sorry, I have no time to investigate your example. I had a quick look but it was not obvious to me how to apply it or what the issue is.
